src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.ActivatorCache.cs

```diff
@@ -0,0 +1,130 @@
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    internal sealed partial class RuntimeType
+    {
+        /// <summary>
+        /// A cache which allows optimizing <see cref="Activator.CreateInstance"/>,
+        /// <see cref="RuntimeType.CreateInstanceDefaultCtor"/>, and related APIs.
+        /// </summary>
+        private sealed unsafe class ActivatorCache
+        {
+            // The managed calli to the newobj allocator, plus its first argument (MethodTable*).
+            // In the case of the COM allocator, first arg is ComClassFactory*, not MethodTable*.
+            private readonly delegate*<void*, object?> _pfnAllocator;
+            private readonly void* _allocatorFirstArg;
+
+            // The managed calli to the parameterless ctor, taking "this" (as object) as its first argument.
+            private readonly delegate*<object?, void> _pfnCtor;
+            private readonly bool _ctorIsPublic;
+
+#if DEBUG
+            private readonly RuntimeType _originalRuntimeType;
+#endif
+
+            internal ActivatorCache(RuntimeType rt)
+            {
+                Debug.Assert(rt != null);
+
+#if DEBUG
+                _originalRuntimeType = rt;
+#endif
+
+                // The check below is redundant since these same checks are performed at the
+                // unmanaged layer, but this call will throw slightly different exceptions
+                // than the unmanaged layer, and callers might be dependent on this.
+
+                rt.CreateInstanceCheckThis();
+
+                try
+                {
+                    RuntimeTypeHandle.GetActivationInfo(rt,
+                        out _pfnAllocator!, out _allocatorFirstArg,
+                        out _pfnCtor!, out _ctorIsPublic);
+                }
+                catch (Exception ex)
+                {
+                    // Exception messages coming from the runtime won't include
+                    // the type name. Let's include it here to improve the
+                    // debugging experience for our callers.
+
+                    string friendlyMessage = SR.Format(SR.Activator_CannotCreateInstance, rt, ex.Message);
+                    switch (ex)
+                    {
+                        case ArgumentException: throw new ArgumentException(friendlyMessage);
+                        case PlatformNotSupportedException: throw new PlatformNotSupportedException(friendlyMessage);
+                        case NotSupportedException: throw new NotSupportedException(friendlyMessage);
+                        case MethodAccessException: throw new MethodAccessException(friendlyMessage);
+                        case MissingMethodException: throw new MissingMethodException(friendlyMessage);
+                        case MemberAccessException: throw new MemberAccessException(friendlyMessage);
+                    }
+
+                    throw; // can't make a friendlier message, rethrow original exception
+                }
```

Comment:
Is this better than, say:
```C#
catch (ArgumentException e) { throw new ArgumentException(CreateFriendlyMessage(rt, e)); }
catch (PlatformNotSupportedException e) { throw new PlatformNotSupportedException(CreateFriendlyMessage(rt, e)); }
... // etc.
```
?  I realize the above is probably a bit more IL, but it also means you won't catch and rethrow if the exception didn't match one of the special-cased types.  I don't know how common that will be.

Reply:
Not ignoring this, but leaving this thread marked unresolved for now because I think we can address it as part of https://github.com/dotnet/runtime/issues/36194#issuecomment-736705053.